### PR TITLE
fix for referencing merged list/map entries

### PR DESCRIPTION
**Pull request by mandelsoft**
Unfortunately I found a bug in my new merge handling in combination with reference evaluation.
I added the tests for those cases and provide the fix.

Entries overridden by an inline merge should be referenced with the
overridden value, but so far the original value is used

```yaml
---
foo:
  <<: (( merge ))
  bar: 42
ref:
  bar: (( foo.bar ))
```
merged with
```yaml
---
foo:
  bob: added!
  bar: overridden
```
should resolve ref to `overridden` not `42`.

problem occurs for all variants of inline merges

### CODE REVIEW

**Comment by mandelsoft** [comment]
May be there is another case hat should be discussed.
I played a little bit with the redirecting merge.

Let's have a look at the following scenario:
```yaml
foo: 
  <<: (( merge alt ))
  bar:
    <<: (( merge ))
    alice: alice
````
merged with
```yaml
foo: 
  alice: not merged!
alt: 
  bar:
    alice: overwritten
    bob: added!
```

What should be a useful behaviour of the nested unqualified merge (in `foo.bar`), if there is an outer redirecting merge? The actual behaviour does not take the outer merge into account.
What I expected without further thinking about the implementation was that missing fields from the stub are inserted. But it didn't happen, because the merge implictly uses the acutal path `meta.bar` not the implied redirected path `alt.bar`.
Basically I could explicitly  write `merge alt.bar`. But my impression is that intuitively the expectation would be to implicitly take the redirection into account and use the implied redirection path.

This could easily be changed, because all the information is there, but I want to discuss this with you, before adding this commit.

Thanks
Uwe

**Comment by mandelsoft** [comment]
In branch [lambda](https://github.com/mandelsoft/spiff/blob/lambda/README.md) there are the additional functions `split` and `trim` and the support of lambda expressions to define additional functions as part of the yaml document. May be you want to have a look. The commits are based on the ones here, but I don't want to bundle new functionality with the fixes provided by this PR.

**Comment by Amit-PivotalLabs** [comment]
In the case of

```yaml
foo: 
  <<: (( merge alt ))
  bar:
    <<: (( merge ))
    alice: alice
````
merged with
```yaml
foo: 
  alice: not merged!
alt: 
  bar:
    alice: overwritten
    bob: added!
```

I'd probably expect the output to be

```yaml
foo:
  bar:
    alice: overwritten
    bob: added!
```

**Comment by mandelsoft** [comment]
Yepp, this is exactly what I would expect, too. I added the fix for this to this PR in 00960dd

**Comment by robdimsdale** [comment]
There was a couple of small typos, and some tabs in the example in the README which we removed as tabs are not valid yaml. I fixed these in a commit.

Thanks for the PR.
